### PR TITLE
handle typed nil on field

### PR DESCRIPTION
**Pull request by vvakame**
it raised panic. I fixed it.

so, I want to add tests about `Y, Z []int` for `nil` and `([]int)(nil)`, but its not compatible to `json.Marshal`.

### CODE REVIEW

File: funcr/funcr_test.go [hunk 1]
```diff
@@ -95,6 +95,11 @@ func TestPretty(t *testing.T) {
 			{"nine", "three"},
 			{"seven", "six"},
 		},
+		struct {
+			X *int
+		}{
+			X: (*int)(nil), // typed nil
```
Comment (pohly) [inline funcr/funcr_test.go:101]: Why not just `nil` or omit the entire initialization?

`X: nil` may be useful to point out that the key aspect of the test case is about a pointer being nil, even though it is not required. `X: (*int)(nil)` just seems unnecessarily complex.

File: funcr/funcr_test.go [hunk 2]
```diff
@@ -95,6 +95,11 @@ func TestPretty(t *testing.T) {
...
+		},
```
Comment (pohly) [inline funcr/funcr_test.go:102]: It would be useful to have more test cases:
- non-nil pointer (to ensure that it works)
- nil interface
- non-nil interface

File: funcr/funcr_test.go (hunk 1 above)
Comment (vvakame) [reply funcr/funcr_test.go]: because, some library uses typed nil on field.
I agree with you that just `nil` doesn't bring any trouble... 😿 
e.g. https://github.com/vektah/gqlparser/blob/5a9c950423813b706a1783f22955c73e6db7450a/parser/schema.go#L201-L205

File: funcr/funcr_test.go (hunk 2 above)
Comment (vvakame) [reply funcr/funcr_test.go:108]: added 👍

File: funcr/funcr_test.go (hunk 1 above)
Comment (pohly) [reply funcr/funcr_test.go]: The code snippet doesn't do anything with `nil`?

My point is: `nil` and `(*int)(nil)` are equivalent initializers for `int *`, so why use the more complex syntax?

File: funcr/funcr_test.go (hunk 2 above)
Comment (pohly) [reply funcr/funcr_test.go:108]: Thanks. You can squash directly, for such a simple change it's easy to remember what changed and for more complex ones the Github diff for force-pushes works well.

File: funcr/funcr_test.go (hunk 1 above)
Comment (vvakame) [reply funcr/funcr_test.go]: ah, make sense.
typed nil is not exactly same with nil.
https://play.golang.org/p/oOaPEgllth3

so, if field value is simply `nil`, value are trapped here. print `null`.
https://github.com/go-logr/logr/blob/5e3dfaa87e750be4ffe6702bb1022d8df301f19d/funcr/funcr.go#L210-L213 .

but if field value is typed nil, it causes a panic.
This PR's test case will be panic without my patch

<details>

```
$ go test ./...
ok  	github.com/go-logr/logr	(cached)
ok  	github.com/go-logr/logr/benchmark	(cached) [no tests to run]
?   	github.com/go-logr/logr/examples	[no test files]
--- FAIL: TestPretty (0.00s)
panic: reflect: call of reflect.Value.Interface on zero Value [recovered]
	panic: reflect: call of reflect.Value.Interface on zero Value

goroutine 18 [running]:
testing.tRunner.func1.2({0x1112be0, 0xc0000b85e8})
	/usr/local/Cellar/go/1.17/libexec/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/Cellar/go/1.17/libexec/src/testing/testing.go:1212 +0x218
panic({0x1112be0, 0xc0000b85e8})
	/usr/local/Cellar/go/1.17/libexec/src/runtime/panic.go:1038 +0x215
reflect.valueInterface({0x0, 0x0, 0x17}, 0x89)
	/usr/local/Cellar/go/1.17/libexec/src/reflect/value.go:1356 +0x10e
reflect.Value.Interface(...)
	/usr/local/Cellar/go/1.17/libexec/src/reflect/value.go:1351
github.com/go-logr/logr/funcr.prettyWithFlags({0x11092a0, 0x0}, 0x0)
	/tmp/logr/funcr/funcr.go:290 +0x945
github.com/go-logr/logr/funcr.pretty(...)
	/tmp/logr/funcr/funcr.go:162
github.com/go-logr/logr/funcr.prettyWithFlags({0x1121ac0, 0xc00009a390}, 0x0)
	/tmp/logr/funcr/funcr.go:255 +0xcec
github.com/go-logr/logr/funcr.pretty(...)
	/tmp/logr/funcr/funcr.go:162
github.com/go-logr/logr/funcr.TestPretty(0xc000083380)
	/tmp/logr/funcr/funcr_test.go:112 +0x79f
testing.tRunner(0xc000083380, 0x113a728)
	/usr/local/Cellar/go/1.17/libexec/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/Cellar/go/1.17/libexec/src/testing/testing.go:1306 +0x35a
FAIL	github.com/go-logr/logr/funcr	0.643s
?   	github.com/go-logr/logr/funcr/example	[no test files]
ok  	github.com/go-logr/logr/testing	(cached)
FAIL
```

</details>

File: funcr/funcr_test.go (hunk 2 above)
Comment (vvakame) [reply funcr/funcr_test.go:108]: did it!

File: funcr/funcr_test.go (hunk 1 above)
Comment (pohly) [reply funcr/funcr_test.go]: I understand that `nil` and `(*int)(nil)` are different constants. But here they are used to initialize an `int *` struct member. The value of that struct member is going to be the same regardless of how it got initialized, right? So why not use the simpler `nil`?

All of the following test cases fail the same way:
```
		struct{
			X *int
		}{},
		struct{
			X *int
		}{
			X: nil,
		},
		struct{
			X *int
		}{
			X: (* int)(nil),
		},
```

```
--- FAIL: TestPretty (0.00s)
panic: reflect: call of reflect.Value.Interface on zero Value [recovered]
	panic: reflect: call of reflect.Value.Interface on zero Value

goroutine 18 [running]:
...
github.com/go-logr/logr/funcr.prettyWithFlags({0x509a00, 0x0}, 0x0)
	/nvme/gopath/src/github.com/go-logr/logr/funcr/funcr.go:290 +0x945
```

Comment (vvakame) [reply funcr/funcr_test.go]: ah, okay. I'm sorry. I'm wrong.
I thought there was a difference in the initial value, but I understood it in your explanation. thanks! 😸 
I updated this PR.

**Review by pohly — APPROVED** [review]
Thanks for the fix.
